{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# mtDNA analysis\n",
    "## Defining of different haplogroups and how this was done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From aligned fasta including HGDP individuals, subset HGDP set to wanted countries\n",
    "We also add to the HGDP individual name the ethnic group and haplogroup call."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1068,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">HGDP00469\n",
      "['Brazil', 'Italy', 'Israel_(Central)', 'Cambodia', 'Democratic_Republic_of_Congo', 'Bougainville', 'Nigeria', 'Israel_(Negev)', 'Italy_(Bergamo)', 'Orkney_Islands', 'South_Africa', 'China', 'Russia_Caucasus', 'Siberia', 'Israel_(Carmel)', 'Pakistan', 'Country', 'Namibia', 'Kenya', 'France', 'Russia', 'Central_African_Republic', 'Mexico', 'New_Guinea', 'Senegal', 'Colombia', 'Japan', 'Algeria_(Mzab)']\n",
      "[45, 36, 48, 11, 15, 22, 25, 49, 14, 16, 8, 184, 17, 25, 48, 198, 1, 7, 12, 53, 25, 36, 50, 17, 24, 13, 31, 29]\n",
      "240\n"
     ]
    }
   ],
   "source": [
    "##Filter for certain HGDP data -- like Asians only?\n",
    "import bioinformatics as bio\n",
    "import collections\n",
    "pD=\"/mnt/solexa/mel_yang/bronzeage/mtdna/\"\n",
    "hgdpinfo=open(pD+\"HGDP_info.txt\",'r')\n",
    "fastafile=open(pD+\"Shang+HGDPaligned_manual.fasta\")\n",
    "\n",
    "myids,myseqs=bio.parsefasta(pD+\"Shang+HGDPaligned_manual.fasta\")\n",
    "print myids[50]\n",
    "\n",
    "myloc=collections.defaultdict(list)\n",
    "myhaps=collections.defaultdict(list)\n",
    "mydata=[]\n",
    "myinds=[]\n",
    "for ind,line in enumerate(hgdpinfo):\n",
    "    x = line.split('\\t')\n",
    "    myloc[x[3]].append(x[0])\n",
    "    myhaps[x[3]].append(x[2]+\"|\"+x[6])\n",
    "    #mydata.append(x[2]+\"|\"+x[6])\n",
    "    #myinds.append(x[0])\n",
    "    #if ind==10:break\n",
    "print myloc.keys()\n",
    "print [len(myloc[i]) for i in myloc.keys()]\n",
    "wantedids=myloc[\"China\"]+myloc[\"Japan\"]+myloc[\"Nigeria\"] #myinds #myloc[\"Cambodia\"]+myloc[\"China\"]+myloc[\"Japan\"]\n",
    "wantedhaps=myhaps[\"China\"]+myhaps[\"Japan\"]+myhaps[\"Nigeria\"] #mydata #myhaps[\"Cambodia\"]+myhaps[\"China\"]+myhaps[\"Japan\"]\n",
    "print len(wantedids)\n",
    "\n",
    "newids,newseqs=[],[]\n",
    "for ind,myid in enumerate(myids):\n",
    "    if myid[1:] in wantedids or \"HGDP\" not in myid: \n",
    "        if myid[1:] in wantedids: \n",
    "            newids.append(myid+\"|\"+wantedhaps[wantedids.index(myid[1:])])\n",
    "        else: newids.append(myid)\n",
    "        newseqs.append(myseqs[ind])\n",
    "    else:\n",
    "        continue\n",
    "        \n",
    "#anno_Shang+HGDPaligned_manual.fasta\n",
    "#Cambodia_China_Japan_Shang+HGDPaligned_manual.fasta\n",
    "newfile=open(pD+\"Nigeria_China_Japan_Shang+HGDPaligned_manual.fasta\",'w') \n",
    "for ind,myseq in enumerate(newseqs):\n",
    "    newfile.write('%s\\n' % (newids[ind]))\n",
    "    for ind2,bp in enumerate(myseq):\n",
    "        if ind2%70!=0: continue\n",
    "        newfile.write(myseq[ind2:(ind2+70)]+'\\n')\n",
    "newfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1057,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name '__version__' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1057-cfa3b36ce590>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mprint\u001b[0m \u001b[0m__version__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name '__version__' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List of defining mtDNA for different haplogroups\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "L123456=\"C146T  C182T  T4312C  T10664C  C10915T  A11914G  G13276A  G16230A\"\n",
    "L23456=\"C152T  A2758G  C2885T  G7146A  T8468C\"\n",
    "L2346=\"C195T  A247G  A825t  T8655C  A10688G  C10810T  G13105A  T13506C  G15301A  A16129G  T16187C  C16189T\"\n",
    "L346=\"G4104A  A7521G\"\n",
    "L34=\"T182C!  T3594C  T7256C  T13650C  T16278C\"\n",
    "L3=\"A769G  A1018G  C16311T\"\n",
    "\n",
    "N=\"G8701A  C9540T  G10398A  C10873T  A15301G!\"\n",
    "R=\" T12705C  T16223C\"\n",
    "\n",
    "M=\"T489C  C10400T  T14783C  G15043A\"\n",
    "M80D=\"C4883T\"\n",
    "D=\"C5178a  T16362C\"\n",
    "D4=\"G3010A  C8414T  C14668T\"\n",
    "D5=\"T16189C! C150T  T1107C  A5301G  A10397G\"\n",
    "D6=\"T16189C! G709A  G1719A  A3714G  A12654G  T16311C!\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script for finding unique SNPs in one sequence relative to another\n",
    "Needs aligned and rCRS in fasta file, and positions should correspond to rCRS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 982,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16569\n",
      "\n",
      "##ref|hg|ID:R2062|XS|R2055\n",
      "##Missing sites:2\n",
      "##Number different in Control (16024-576):2\n",
      "##Number different in Coding (577-16023):0\n",
      "\n",
      "\tPos\thg\tR2055\n",
      "Control\t3\tN\tT\n",
      "Control\t310\tC\tT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import bioinformatics as bio\n",
    "\n",
    "pD=\"/mnt/solexa/mel_yang/bronzeage/mtdna/\"\n",
    "myids,myseqs=bio.parsefasta(pD+\"aligned_Shang+rCRS+refseqs_manual.fasta\") \n",
    "\n",
    "ref,myid=(\"JQ704974\",\"R2062\") \n",
    "ref,myid=(\"JQ704974\",\"D5604\") \n",
    "ref,myid=(\"KC733262\",\"R2084\") \n",
    "ref,myid=(\"JF824906\",\"D5606\") \n",
    "ref,myid=(\"R2062\",\"R2055\")\n",
    "refind=[ind for ind,i in enumerate(myids) if ref in i]\n",
    "myidind=[ind for ind,i in enumerate(myids) if myid in i]\n",
    "\n",
    "refseq=myseqs[refind[0]]\n",
    "myseq=myseqs[myidind[0]]\n",
    "print len(myseqs[0])\n",
    "missing=0\n",
    "numcontrol,numcoding=0,0\n",
    "mystr=\"\\t%s\\t%s\\t%s\\n\" % (\"Pos\",\"hg\",myid)\n",
    "for pos,i in enumerate(myseq):\n",
    "    if pos < 576 or pos > 16023: addon=\"Control\"\n",
    "    else: addon=\"\" #myid\n",
    "    if refseq[pos] != myseq[pos]: \n",
    "        if myseq[pos] in [\"N\",\"-\"]: missing+=1\n",
    "        else: \n",
    "            if addon==\"Control\": numcontrol+=1\n",
    "            else: numcoding+=1\n",
    "            mystr += \"%s\\t%i\\t%s\\t%s\\n\" % (addon,pos+1,refseq[pos],myseq[pos])\n",
    "print\n",
    "print \"##ref|hg|ID:%s|%s\" % (myids[refind[0]][1:],myid) \n",
    "print \"##Missing sites:%i\" % missing\n",
    "print \"##Number different in Control (16024-576):%i\" % numcontrol\n",
    "print \"##Number different in Coding (577-16023):%i\" % numcoding\n",
    "print\n",
    "print mystr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "import bioinformatics as bio\n",
    "import collections\n",
    "pD=\"/mnt/solexa/mel_yang/bronzeage/mtdna/\"\n",
    "\n",
    "mysites={\"XSQ\":[\"D5603\",\"D5602\",\"D5601\",\"D1967\",\"D1962\",\"D1968\",\"D1965\",\"D1960\",\"D1961\"],\n",
    "        \"LZ\":[\"R2038\",\"R2039\",\"R2024\",\"R2087\",'R2023'],\n",
    "        \"GDM\":[\"D5604\",\"D5605\",\"D5608\",\"D5606\",\"D5607\"],\n",
    "        \"YJ\":[\"R2084\"],\n",
    "        \"XS\":[\"R2086\",\"R2055\",\"R2062\",\"R2123\",\"R2090\"],\n",
    "        \"XHC\":[\"D5611\",\"D5613\",\"D5612\"]}\n",
    "sites,ids=[],[]\n",
    "for i in mysites:\n",
    "    sites+=[i]*len(mysites[i])\n",
    "    ids+=mysites[i]\n",
    "\n",
    "myids,myseqs=bio.parsefasta(pD+\"aligned_pdEAS+rCRS+Shang.fasta\")\n",
    "#myids,myseqs=bio.parsefasta(pD+\"Shang.fasta\")\n",
    "#myids,myseqs=bio.parsefasta(pD+\"Shang+RSRS_MAF.fasta\")\n",
    "#myids,myseqs=bio.parsefasta(pD+\"Shang_rCRS_muscle_haplogrep.fasta\")\n",
    "\n",
    "refind= [ind for ind,i in enumerate(myids) if \"012920\" in i][0]\n",
    "refseq=myseqs[refind]\n",
    "\n",
    "myseqs=myseqs[:refind]+myseqs[(refind+1):]\n",
    "myids=myids[:refind]+myids[(refind+1):]\n",
    "#myseqs=[myseqs[ind] for ind,i in enumerate(myids) if \"HGDP\" not in i]\n",
    "#myids=[i for ind,i in enumerate(myids) if \"HGDP\" not in i]\n",
    "myseqs=[myseqs[ind] for ind,i in enumerate(myids)]\n",
    "myids=[i for ind,i in enumerate(myids)]\n",
    "\n",
    "newids=[]\n",
    "for myid in myids:\n",
    "    if 'min_cov5' in myid:\n",
    "        if 'Split' in myid: newids.append(myid[1:].split('|')[0].split('_')[1])\n",
    "        else: newids.append(myid[1:].split('.')[0])\n",
    "    else: newids.append(myid)\n",
    "#newids=[myid[1:].split('|')[0].split('_')[1] if 'Split' in myid else myid[1:].split('.')[0]  for myid in myids ]\n",
    "badids=['D5612','D1960','R2123','R2090','D5613','R2023'] \n",
    "#Id #Missing\n",
    "#D5612 6891\n",
    "#D5613 2948\n",
    "#R2023 2396\n",
    "#R2123 2007\n",
    "#R2090 1453\n",
    "#D1960 1167\n",
    "#for i in newids: print i\n",
    "filtseqs=[i for ind,i in enumerate(myseqs) if newids[ind] not in badids]\n",
    "filtids = [i for ind,i in enumerate(newids) if newids[ind] not in badids]\n",
    "\n",
    "newbases=['' for i in myseqs]\n",
    "delpos=[]\n",
    "num=0\n",
    "for posind,base in enumerate(refseq): ##posind==position on sequence\n",
    "    if base == '-': delpos.append(posind); continue\n",
    "    num+=1\n",
    "    mybases=[]\n",
    "    for myind,i in enumerate(myseqs): ##myind==each mtDNA sequence (each individual)\n",
    "        if num==3107: newbases[myind]+='N' ##3107 deletion in rCRS, artifact?\n",
    "        elif num==16519: newbases[myind]+='N' ##hotspot\n",
    "        elif num in range(515,523): newbases[myind]+='N' #AC indels repeats\n",
    "        elif num in range(16182,16184): newbases[myind]+='N'\n",
    "        else: newbases[myind]+=i[posind]\n",
    "        mybases.append(i[posind])\n",
    "    #if num==10: break\n",
    "\n",
    "    uniqbases=collections.Counter(mybases)\n",
    "    #if len(uniqbases.keys())>1:print base, num,uniqbases\n",
    "\n",
    "newfile=open(pD+\"Shang+HGDPaligned_manual.fasta\",'w')\n",
    "for ind,myseq in enumerate(newbases):\n",
    "    if \"HGDP\" not in newids[ind]: newfile.write('>%s|%s\\n' % (newids[ind],sites[ids.index(newids[ind])]))\n",
    "    else: newfile.write(newids[ind] + '\\n')\n",
    "    for ind2,bp in enumerate(myseq):\n",
    "        if ind2%70!=0: continue\n",
    "        newfile.write(myseq[ind2:(ind2+70)]+'\\n')\n",
    "newfile.close()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get rCRS aligned fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 938,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "import bioinformatics as bio\n",
    "import collections\n",
    "pD=\"/mnt/solexa/mel_yang/bronzeage/mtdna/\"\n",
    "\n",
    "mysites={\"XSQ\":[\"D5603\",\"D5602\",\"D5601\",\"D1967\",\"D1962\",\"D1968\",\"D1965\",\"D1960\",\"D1961\"],\n",
    "        \"LZ\":[\"R2038\",\"R2039\",\"R2024\",\"R2087\",'R2023'],\n",
    "        \"GDM\":[\"D5604\",\"D5605\",\"D5608\",\"D5606\",\"D5607\"],\n",
    "        \"YJ\":[\"R2084\"],\n",
    "        \"XS\":[\"R2086\",\"R2055\",\"R2062\",\"R2123\",\"R2090\"],\n",
    "        \"XHC\":[\"D5611\",\"D5613\",\"D5612\"]}\n",
    "sites,ids=[],[]\n",
    "for i in mysites:\n",
    "    sites+=[i]*len(mysites[i])\n",
    "    ids+=mysites[i]\n",
    "\n",
    "myids,myseqs=bio.parsefasta(pD+\"aligned_Shang+rCRS+refseqs.fasta\")\n",
    "\n",
    "refind= [ind for ind,i in enumerate(myids) if \"012920\" in i][0]\n",
    "refseq=myseqs[refind]\n",
    "\n",
    "myseqs=myseqs[:refind]+myseqs[(refind+1):]\n",
    "myids=myids[:refind]+myids[(refind+1):]\n",
    "#myseqs=[myseqs[ind] for ind,i in enumerate(myids) if \"HGDP\" not in i]\n",
    "#myids=[i for ind,i in enumerate(myids) if \"HGDP\" not in i]\n",
    "myseqs=[myseqs[ind] for ind,i in enumerate(myids)]\n",
    "myids=[i for ind,i in enumerate(myids)]\n",
    "\n",
    "newids=[]\n",
    "for myid in myids:\n",
    "    if 'min_cov5' in myid:\n",
    "        if 'Split' in myid: newids.append(myid[1:].split('|')[0].split('_')[1])\n",
    "        else: newids.append(myid[1:].split('.')[0])\n",
    "    else: newids.append(myid)\n",
    "\n",
    "badids=['D5612','D1960','R2123','R2090','D5613','R2023'] \n",
    "\n",
    "filtseqs=[i for ind,i in enumerate(myseqs) if newids[ind] not in badids]\n",
    "filtids = [i for ind,i in enumerate(newids) if newids[ind] not in badids]\n",
    "\n",
    "newbases=['' for i in myseqs]\n",
    "delpos=[]\n",
    "num=0\n",
    "for posind,base in enumerate(refseq): ##posind==position on sequence\n",
    "    if base == '-': delpos.append(posind); continue\n",
    "    num+=1\n",
    "    mybases=[]\n",
    "    for myind,i in enumerate(myseqs): ##myind==each mtDNA sequence (each individual)\n",
    "        if num==3107: newbases[myind]+='N' ##3107 deletion in rCRS, artifact?\n",
    "        elif num==16519: newbases[myind]+='N' ##hotspot\n",
    "        elif num in range(515,523): newbases[myind]+='N' #AC indels repeats\n",
    "        elif num in range(16182,16184): newbases[myind]+='N'\n",
    "        else: newbases[myind]+=i[posind]\n",
    "        mybases.append(i[posind])\n",
    "    #if num==10: break\n",
    "\n",
    "    uniqbases=collections.Counter(mybases)\n",
    "    #if len(uniqbases.keys())>1:print base, num,uniqbases\n",
    "\n",
    "newfile=open(pD+\"aligned_Shang+rCRS+refseqs_manual.fasta\",'w')\n",
    "for ind,myseq in enumerate(newbases):\n",
    "    \n",
    "    if newids[ind] in ids: newfile.write('>%s|%s\\n' % (newids[ind],sites[ids.index(newids[ind])]))\n",
    "    else: newfile.write(newids[ind] + '\\n')\n",
    "    for ind2,bp in enumerate(myseq):\n",
    "        if ind2%70!=0: continue\n",
    "        newfile.write(myseq[ind2:(ind2+70)]+'\\n')\n",
    "newfile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indicate number of missing sites per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D5612 6890\n",
      "D5613 2955\n",
      "R2023 2404\n",
      "R2123 2015\n",
      "R2090 1452\n",
      "D1960 1173\n",
      "D5611 287\n",
      "D1961 243\n",
      "D1962 73\n",
      "D5602 58\n",
      "D5607 47\n",
      "D1967 35\n",
      "R2055 16\n",
      "R2062 17\n",
      "R2084 12\n",
      "R2038 27\n",
      "R2039 12\n",
      "D5604 16\n",
      "R2024 23\n",
      "D1968 29\n",
      "D1965 40\n",
      "D5605 15\n",
      "D5603 20\n",
      "R2086 15\n",
      "R2087 13\n",
      "D5601 21\n",
      "D5608 14\n",
      "D5606 18\n"
     ]
    }
   ],
   "source": [
    "for ind,i in enumerate(myseqs):\n",
    "    print newids[ind],sum([1 for bp in i if bp in ['-','N']])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['XHC', 'XHC', 'XHC', 'YJ', 'XSQ', 'XSQ', 'XSQ', 'XSQ', 'XSQ', 'XSQ', 'XSQ', 'XSQ', 'XSQ', 'XS', 'XS', 'XS', 'XS', 'XS', 'LZ', 'LZ', 'LZ', 'LZ', 'LZ', 'GDM', 'GDM', 'GDM', 'GDM', 'GDM']\n",
      "['D5611', 'D5613', 'D5612', 'R2084', 'D5603', 'D5602', 'D5601', 'D1967', 'D1962', 'D1968', 'D1965', 'D1960', 'D1961', 'R2086', 'R2055', 'R2062', 'R2123', 'R2090', 'R2038', 'R2039', 'R2024', 'R2087', 'R2023', 'D5604', 'D5605', 'D5608', 'D5606', 'D5607']\n"
     ]
    }
   ],
   "source": [
    "mysites={\"XSQ\":[\"D5603\",\"D5602\",\"D5601\",\"D1967\",\"D1962\",\"D1968\",\"D1965\",\"D1960\",\"D1961\"],\n",
    "        \"LZ\":[\"R2038\",\"R2039\",\"R2024\",\"R2087\",'R2023'],\n",
    "        \"GDM\":[\"D5604\",\"D5605\",\"D5608\",\"D5606\",\"D5607\"],\n",
    "        \"YJ\":[\"R2084\"],\n",
    "        \"XS\":[\"R2086\",\"R2055\",\"R2062\",\"R2123\",\"R2090\"],\n",
    "        \"XHC\":[\"D5611\",\"D5613\",\"D5612\"]}\n",
    "sites,ids=[],[]\n",
    "for i in mysites:\n",
    "    sites+=[i]*len(mysites[i])\n",
    "    ids+=mysites[i]\n",
    "print sites\n",
    "print ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test what bases are at a particular site--several lists for haplogroup defining SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1062,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[8701, 9540, 10398, 10873, 15301, 489, 10400, 14783, 15043]\n",
      ">D5602|XSQ ----- ['N', 'N', 'N', 'N', 'N', 'N', 'N', 'N', 'N'] ----\n"
     ]
    }
   ],
   "source": [
    "import bioinformatics as bio\n",
    "\n",
    "pD=\"/mnt/solexa/mel_yang/bronzeage/mtdna/\"\n",
    "#myids,myseqs=bio.parsefasta(pD+\"Shang_HGDPaligned_manual.fasta\") \n",
    "#myids,myseqs=bio.parsefasta(pD+\"Shang_manual.fasta\")\n",
    "myids,myseqs=bio.parsefasta(pD+\"aligned_Shang+rCRS+refseqs_manual.fasta\") \n",
    "\n",
    "\n",
    "L3=[769,1018,16311] ##A769G A1018G C16311T\n",
    "N=[8701,9540,10398,10873,15301] #G8701A C9540T G10398A C10873T A15301G!\n",
    "R=[12705,16223] #T12705C T16223C\n",
    "A13=[3504] #T3504C\n",
    "A=[235,663,1736,4248,4824,8794,16290,16319] #A235G A663G A1736G T4248C A4824G C8794T C16290T G16319A\n",
    "Aplus=[152,16362]\n",
    "A5=[8563,11536]\n",
    "A10=[5393,7468,9948,10094,16227,16311]\n",
    "A15=[207,263,8459,14067]\n",
    "A15a=[11084]\n",
    "B45=range(8281,8290)\n",
    "B4=[16217] #T16217C\n",
    "B4i=[6302] #A6302G\n",
    "R9=[3970,13928,16304] #C3970T G13928c T16304C\n",
    "B5=[709,8584,9950,10398,16140]\n",
    "\n",
    "JT=[11251,15452,16126] #A11251G C15452a T16126C\n",
    "P=[15607] ##A15607G\n",
    "F=[249,6392,10310] #A249d T6392C G10310A, from R9\n",
    "F1=[6962,10609,12406,12882] #G6962A T10609C G12406A C12882T\n",
    "F1d=[146,1734,5628,7738,15402] #T146C! C1734T T5628C T7738C C15402T\n",
    "R0=[73,11719] #G73A A11719G\n",
    "HV=[14766] #T14766C\n",
    "H=[2706,7028] #G2706A T7028C\n",
    "\n",
    "R1=[1391,16311] #T1391C T16311C!\n",
    "R2=[152,4769,7657,8473,9932,10685,12654,14305,16071] #T152C! G4769A T7657C T8473C G9932A G10685A A12654G G14305A C16071T\n",
    "R2JT=[4216] #T4216C\n",
    "R12_21=[10398,11404,16295] #A10398G! A11404G C16295T\n",
    "R5=[8594,10754,14544,16304,16524] #T8594C A10754G G14544A T16304C A16524G\n",
    "R6=[195,12285,16266,16362] #(T195C!) T12285C (C16266T) T16362C\n",
    "R7=[13105,16319,16362] #A13105G! G16319A T16362C\n",
    "R8=[195,2755,3384,7759,9449,13215] # T195C! A2755G A3384G T7759C C9449T T13215C\n",
    "R9=[3970,13928,16304] #C3970T G13928c T16304C\n",
    "Rplus=[16189]\n",
    "\n",
    "R14=[56,182,204,234,573,4002,7022,8860,9128,10172,11321,11560,12612,13281,14002,14800,15394,15458,16187,16288,16304,16362] #A56t C182T!! T204C A234G 573.XC T4002C T7022g G8860A T9128C G10172A A11321G A11560G A12612G T13281C A14002G C14800T T15394C T15458C C16187T! T16288C T16304g T16362C\n",
    "R22=[152,329,1040,6770,10316,10609,13359,16249,16288,16304] #T152C! G329A T1040C A6770G A10316G T10609C G13359A T16249C T16288C T16304C\n",
    "R23=[195,634,6216,7007,7158,7388,7822,8614,10987,11914,12151,12459,13885,14687,15412,16256,16290,16465] #T195C! T634a T6216C C7007T A7158G A7388G A7822G T8614C A10987G G11914A! A12151G A12459G C13885a A14687G T15412C C16256T C16290T C16465T\n",
    "R24=[146,3565,4260,9759,11710,12483,16051,16185,16193] #T146C! A3565G C4260T C9759T C11710T C12483T A16051G C16185T C16193d\n",
    "R30=[8584] #G8584A\n",
    "R31=[15884,16362] #G15884A (T16362C)\n",
    "R32=B45+[152,3816,4820,8464,8743,8784,8887,9116,12771,15712,16145,16185,16239,16311,16325] #T152C! A3816G G4820A 8281-8289d C8464T G8743A A8784G A8887G T9116C G12771A A15712G G16145A C16185T C16239T T16311C! T16325C\n",
    "U=[11467,12308,12372] #A11467G A12308G G12372A\n",
    "M=[489,10400,14783,15043] #T489C C10400T T14783C G15043A\n",
    "M2=[447,1780,11083,15670,16274]\n",
    "M3=[482,16126]\n",
    "M5=[1888,16129]\n",
    "M6=[461,5301,5558,10640,14128,16362]\n",
    "M7=[6455,9824] #C6455T T9824C\n",
    "M8=[4715,7196,8584,15487,16298]\n",
    "M80D=[4883] #C4883T\n",
    "M10=[573,709,4140,8793,12549,13152,14502,15040,15071,16311]\n",
    "M11=[215,318,326,1095,6531,7642,8108,9950,11969]\n",
    "M14=[234,4216,6962]\n",
    "M15=[64,183,8167,11002,11629,12662,13659,14284,15412,16147,16193,16249,16362]\n",
    "M17=[930,12973,16209]\n",
    "M22=[710,4639,6776]\n",
    "M25=[150,316,869,1406,3511,7954,8776,11734,13144,13350,13708,16129,16189,16223,16290,16368]\n",
    "M26=[152,5656,9899,11428,11521,12190,13708,13932,14218,14311,15109,15263,16214,16256,16278]\n",
    "M27=[5375,9201,12358]\n",
    "M28=[152,195,1719,6281,6374,10245,15067,16148,16362,16468]\n",
    "M31=[4907,11176,15440,15530]\n",
    "M36=[239,7271,15110]\n",
    "M40=[8925,15721,15954,16463]\n",
    "M44=[146,961,16301]\n",
    "M47=[16129,16166,16189,16287,16319]\n",
    "M48=[199,6366,15900,16225,16234,16390]\n",
    "M49=[3780,16234]\n",
    "M50=[7226,15663,16263]\n",
    "M52=[1462,1598,5460,6020,16275]\n",
    "M58=[3209,5460]\n",
    "M59=[249,9380,10256,11140,14040,16140,16278]\n",
    "M60=[7912,8345]\n",
    "M62_68=[150,4561,7664]\n",
    "M69=[4392,11365]\n",
    "M71=[8718,15458,16271]\n",
    "M72=[1872,14233,15644,15820,16166,16214]\n",
    "M73_79=[14034,16278]\n",
    "M76=[513,3753,7521,10325,12406,14341,16189,16223,16293,16362]\n",
    "M81=[215,4254,6620,13590,16129,16311]\n",
    "M91=[93,3447,4164,15652,16287,16327]\n",
    "\n",
    "\n",
    "D=[5178,16362] #C5178a T16362C, from M80D\n",
    "D4=[3010,8414,14668] #G3010A C8414T C14668T\n",
    "D4a=[152,3206,8473,14979,16129]\n",
    "D4b=[8020]\n",
    "D4d=[3593,5539,6503,7757,9938,12085,13879,16319]\n",
    "D4h=[3336,3644,5048]\n",
    "D4h1=[16174]\n",
    "D4h1c=[152,7181,7673,13914,16311]\n",
    "D4h1c1=[14470,16317]\n",
    "D4k=[195,12092,16362]\n",
    "D4o=[195,10646,16290]\n",
    "D4p=[195,198,15106]\n",
    "D4l=[10427,16368]\n",
    "D4m=[8762,12651]\n",
    "D4n=[3421,16355]\n",
    "D4q=[200,16256,16311]\n",
    "D4s=[199,813,12662]\n",
    "D4t=[263,15901]\n",
    "D1=[2092,16325]\n",
    "D5=[150,1107,5301,10397] #C150T T1107C A5301G A10397G\n",
    "D5ab=[9180] #A9180G\n",
    "D5a=[752,11944,12026] #C752T T11944C A12026G\n",
    "D5a2=[1438,16172] #G1438A T16172C\n",
    "\n",
    "M9=[4491,16362]\n",
    "E=[3027,3705,7598,13626,16390]\n",
    "\n",
    "A3=[2857,8962,9711]\n",
    "A7=[146,8413,10172,15379,16051,16129,16189]\n",
    "A9=[6950,7228,8340,10700,15874]\n",
    "A11=[9650,16293]\n",
    "\n",
    "A2=[146,152,153,8027,12007,16111]\n",
    "A6=[654,3687,13287]\n",
    "A16=[298,2706,6023,11557,14929,16292]\n",
    "A17=[4113,5514,9126,15217]\n",
    "A18=[11465]\n",
    "A19=[199,15436,16311]\n",
    "A20=[14696]\n",
    "A21=[14364]\n",
    "A22=[6378,8393,9123,9128]\n",
    "A24=[449,9524,15172]\n",
    "A25=[16179]\n",
    "A26=[4924,9716,11809,13650,13752,14968,16144]\n",
    "\n",
    "M7c=[146,4850,11665,12091]\n",
    "M7c1=[199,5442,16295]\n",
    "M7c2=[3912,5378,9449,10897,11932,14053,15338,16172,16311]\n",
    "M7c3=[4856,5460,11527,11875,12950,14118]\n",
    "M7c1b=[12561]\n",
    "M7c1b2=[9957,13590]\n",
    "M7c1b2b=[7337,10861]\n",
    "R24=[146,3565,4260,9759,11710,12483,16051,16185,16193]\n",
    "\n",
    "weirddel=range(21,28)\n",
    "\n",
    "B4bdej=[827,15535]\n",
    "B4b=[499,4820,13590]\n",
    "B4b1=[16136]\n",
    "B4b1bc=[16218]\n",
    "B4b1b=[152,1391,3981,5585,13934,14569,16298,16362]\n",
    "B4b1c=[14587]\n",
    "B4b1c1=[7521,9101,11239,11914]\n",
    "N2=[189,709,5046,11674,12414]\n",
    "R9c=[7861,11002,15479]\n",
    "F1=[6962,10609,12406,12882]\n",
    "F2=[1005,1824,7828,10535,10586,12338,13708]\n",
    "F3=[3434,5585,5913,5978,10320,11065,16298,16304,16362]\n",
    "F4=[5263,12630,15670]\n",
    "D6=[709,1719,3714,12654,16311]\n",
    "E=[3027,3705,7598,13626,16390]\n",
    "N9a=[150,5231,12358,12372,16257,16261]\n",
    "N9a13=[16129]\n",
    "N9a1=[4386,12007,16111]\n",
    "N9a1a=[15095]\n",
    "D4c=[16245]\n",
    "D4e=[11215]\n",
    "D4f=[7270]\n",
    "D4g=[13104]\n",
    "D4i=[16294]\n",
    "D4j=[11696]\n",
    "D4j12=[7353,8273,15470]\n",
    "D5c=[151,152,4200,4216,14927,15622,16190]\n",
    "D5c1=[146,3546,13105,16390]\n",
    "D5c1a=[182,217,16223]\n",
    "D5c2=[16311,7129,7669,8580,12810,13984,15737,16316]\n",
    "D6a=[16274]\n",
    "D6a1=[6701,7424,7879,13194,16311]\n",
    "D6a2=[4679,4937,14530,16129,16189,16317]\n",
    "A8a=[16242,64,146]\n",
    "A8a1=[152,6962,8865,12777]\n",
    "F2c2=[761,5978,6156,6570,8152,14133,15418,16067,16271]\n",
    "D4h1c=[152,7181,7673,13914,16311]\n",
    "N3=[146,210,571,1005,3357,6366,6806,7897,8653,8937,10586,14634,16086,16172,16187,16189,16217]\n",
    "N7=[723,6570,11617,13542,14668,15945,16129,16318]\n",
    "N8=[2760,3027,7885,7961,8188,10398,13710,15211,16263,16274,16311,16343]\n",
    "N9=[5417]\n",
    "N10=[750,5581,16172,16362]\n",
    "\n",
    "N13=[146,152,178,195,2831,5704,6896,10101,11020,13263,16066,16261,16290]\n",
    "N14=[1598,1888,4917,5291,8251,8269,11491,14384,15040,15884]\n",
    "N21=[150,337,6752,8701,10583,13437,14560,16193]\n",
    "N22=[150,942,7158,9254,11365,16168,16249]\n",
    "O=[6755,9140,16213]\n",
    "S=[8404]\n",
    "X=[6221,6371,13966,14470,16189,16278]\n",
    "N9a=[150,5231,12358,12372,16257,16261]\n",
    "N9b=[5147,10607,11016,13183,14893,16189]\n",
    "Y=[8392,10398,14178,14693,16126,16223,16231]\n",
    "M1=[195,6446,6680,12403,12950,16129,16189,16249,16311]\n",
    "M20=[152,249,316,3200,3714,4385,4772,7433,8853,9127,9512,10274,10679,11914,12354,14974,15691,16129,16209,16272]\n",
    "M51=[150,4697,4973,9509,14356,14687,15317,16278]\n",
    "M4=[6620,7859,16145,16261]\n",
    "M67=[143,2833,4227,5979,6366,9863,12406,13708,15769,15833,16075,16169]\n",
    "M18=[12498,13135,16318]\n",
    "M38=[4099,15314,15487]\n",
    "M12=[4170,5580,12030,12372,14727,15010,16234,16290]\n",
    "G=[709,4833,5108,16362]\n",
    "M80=[249,709,1352,2695,4197,4823,5492,5894,7939,8429,8775,9254,11732,14118,16134,16189,16325]\n",
    "D4e5=[5021,15106,15184,16274]\n",
    "R0ab=[58,64,2442,16362]\n",
    "A5a=[2156,4655,11647,16187]\n",
    "A5b=[961,965]\n",
    "A5c=[152,16129,16213]\n",
    "A5b1=[1709,16126,16235]\n",
    "A8=[16242]\n",
    "A8a=[64,146]\n",
    "mysets=[(\"D5604\",L3+M+M80D+D+D4),\n",
    "        (\"D5605\",L3+N+R+R9+F+F1+F1d),\n",
    "        (\"D5608\",L3+N+A+A13),\n",
    "        (\"D5606\",L3+N+R+B45+B4+B4i),\n",
    "        (\"D5607\",L3+M+M80D+D+D5+D5ab+D5a),\n",
    "        \n",
    "        (\"R2038\",L3+M+M80D+D+D4+D4j+D4j12),\n",
    "        (\"R2039\",L3+M+M80D+D+D6+D6a),\n",
    "        (\"R2024\",L3+M+M80D+D+D4),\n",
    "        (\"R2087\",L3+N+A+A5+A5b+A5b1),\n",
    "        (\"R2023\",L3+N),\n",
    "        \n",
    "        (\"D5611\",L3+N),\n",
    "        (\"D5613\",L3+N+R),\n",
    "        (\"D5612\",L3+M),\n",
    "        \n",
    "        (\"R2086\",L3+N+A+A8+A8a),\n",
    "        (\"R2055\",L3+M+M80D+D+D5+D5c+D5c2),\n",
    "        (\"R2062\",L3+M+M80D+D+D5+D5c+D5c2),\n",
    "        (\"R2123\",L3+N+R+R9+F+F2+F2c+[761,14133]),\n",
    "        (\"R2090\",L3+M+M80D+D+D5+D5ab+D5a),\n",
    "        \n",
    "        (\"D5603\",L3+N+N9+N9a1),\n",
    "        (\"D5602\",L3+N),\n",
    "        (\"D5601\",L3+N+A+[152,16362]),\n",
    "        (\"D1967\",L3+M+M7+M7c+M7c1+M7c1b+M7c1b2+M7c1b2b),\n",
    "        (\"D1962\",L3+M+M80D+D+D4),\n",
    "        (\"D1968\",L3+N+R+B45+B4+B4bdej+B4b+B4b1+B4b1bc+B4b1c+B4b1c1),\n",
    "        (\"D1965\",L3+N+A+[152,16362]+A15+A15a),\n",
    "        (\"D1960\",L3+N+R),\n",
    "        (\"D1961\",L3+M+M9+E),\n",
    "        (\"R2084\",L3+M+M80D+D+D4+D4h+D4h1+D4h1c)]\n",
    "mysets=[(\"D5602\",N+M)]\n",
    "#myid=\"R2023\"\n",
    "#poses=L3+N+R+B45+B4+B4bdej+B4b+B4b1+B4b1bc+B4b1c+B4b1c1\n",
    "#L3+M+M80D+D+D6+D6a+D6a2\n",
    "for myid,poses in mysets:\n",
    "    poses=[i-1 for i in poses]\n",
    "    print [i+1 for i in poses]\n",
    "    for ind,myseq in enumerate(myseqs):\n",
    "        if 'HGDP' in myids[ind]: continue\n",
    "        subset=[myseq[pos] for pos in poses]\n",
    "        if myid in myids[ind]: \n",
    "            print myids[ind],'-----',subset,'----'\n",
    "            #print myid,len([i for i in subset if i not in '-N']), len(subset)\n",
    "        else: pass #print myids[ind],'\\t',subset\n",
    "    #print [i+1 for i in poses]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
